src/flag_gems/ops/celoss.py

```diff
@@ -0,0 +1,229 @@
+    pid_d = tl.program_id(1)
+    offset_d = pid_d * BLOCK_D + tl.arange(0, BLOCK_D)
+
+    tmp_max = tl.zeros([BLOCK_C, BLOCK_D], dtype=tl.float32)
+    tmp_sum = tl.zeros([BLOCK_C, BLOCK_D], dtype=tl.float32)
+
+    for off in range(0, C, BLOCK_C):
+        offset_c = off + tl.arange(0, BLOCK_C)
+        inp_ptrs = inp_ptr + pid_n * C * D + offset_c[:, None] * D + offset_d[None, :]
+        inp_mask = offset_c[:, None] < C and offset_d[None, :] < D
+        inp = tl.load(inp_ptrs, inp_mask, other=-float("inf")).to(tl.float32)
+        cur_max = tl.maximum(tmp_max, inp)
+        cur_exp = tl.exp(inp - cur_max)
+        tmp_sum = tmp_sum * tl.exp(tmp_max - cur_max) + cur_exp
+        tmp_max = cur_max
+    final_max = tl.max(tmp_max, axis=0)
+    tmp_sum = tmp_sum * tl.exp(tmp_max - final_max[None, :])
+    final_sum = tl.sum(tmp_sum, axis=0)
+
+    _sum = tl.zeros([BLOCK_C, BLOCK_D], dtype=tl.float32)
+    for off in range(0, C, BLOCK_C):
+        offset_c = off + tl.arange(0, BLOCK_C)
+        inp_ptrs = inp_ptr + pid_n * C * D + offset_c[:, None] * D + offset_d[None, :]
+        tgt_ptrs = tgt_ptr + pid_n * C * D + offset_c[:, None] * D + offset_d[None, :]
+        mask = offset_c[:, None] < C and offset_d[None, :] < D
+        w_ptrs = w_ptr + offset_c
+        w_mask = offset_c < C
+        inp = tl.load(inp_ptrs, mask, other=0).to(tl.float32)
+        tgt = tl.load(tgt_ptrs, mask, other=1).to(tl.float32)
+        w = tl.load(w_ptrs, w_mask, other=0)[:, None]
+        log = tl.log(tgt * tl.exp(inp - final_max[None, :]) / final_sum[None, :])
+        _sum -= w * log
+
+    out = tl.sum(_sum, axis=0)
+    out_ptrs = out_ptr + pid_n * D + offset_d
+    tl.store(out_ptrs, out, mask=offset_d < D)
+
+
+class CrossEntropyLoss(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, inp, target, weight, reduction, ignore_index, label_smoothing):
+        logging.debug("GEMS CrossEntropyLoss")
+        # label_smoothing not supported
+
+        shape = list(inp.shape)
+        dim = inp.ndim
+        N = 1 if dim == 1 else shape[0]
+        C = shape[0] if dim == 1 else shape[1]
+        D = inp.numel() // N // C
+        axis = 0 if dim == 1 else 1
+        del shape[axis]
+
+        if weight is None:
+            weight = torch.ones(
+                [
+                    C,
+                ],
+                dtype=inp.dtype,
+                device=inp.device,
+            )
+
+        inp = inp.contiguous()
+        tgt = target.contiguous()
+        weight = weight.contiguous()
+        out = torch.empty(shape, dtype=inp.dtype, device=inp.device)
+        grid = lambda meta: (N, triton.cdiv(D, meta["BLOCK_D"]))
+
+        if target.ndim == dim:
+            # target probabilities
+            with torch.cuda.device(inp.device):
+                celoss_probability_kernel[grid](inp, tgt, weight, out, N, C, D)
+        else:
+            # target indices
+            w_tgt = torch.empty(shape, dtype=inp.dtype, device=inp.device)
```

Comment:
Is this (`w_tgt`) only for the purpose of debugging?

Reply:
reduction mode 'mean' needs it